discovery-provider/src/tasks/entity_manager/delegate.py

```diff
@@ -0,0 +1,127 @@
+import json
+import logging
+from typing import Optional, TypedDict, Union, cast
+
+from src.models.delegates.delegate import Delegate
+from src.tasks.entity_manager.utils import Action, EntityType, ManageEntityParameters
+from src.utils.indexing_errors import EntityMissingRequiredFieldError
+from src.utils.model_nullable_validator import all_required_fields_present
+
+logger = logging.getLogger(__name__)
+
+
+class CreateDelegateMetadata(TypedDict):
+    address: Union[str, None]
+    name: Union[str, None]
+    is_personal_access: Union[bool, None]
+
+
+def get_create_delegate_metadata_from_raw(
+    raw_metadata: Optional[str],
+) -> Optional[CreateDelegateMetadata]:
+    metadata: CreateDelegateMetadata = {
+        "address": None,
+        "name": None,
+        "is_personal_access": None,
+    }
+    if raw_metadata:
+        try:
+            json_metadata = json.loads(raw_metadata)
+            raw_address = json_metadata.get("address", None)
+            if raw_address:
+                metadata["address"] = raw_address.lower()
+            else:
+                metadata["address"] = None
+            metadata["name"] = json_metadata.get("name", None)
+            metadata["is_personal_access"] = json_metadata.get(
+                "is_personal_access", None
+            )
+            return metadata
+        except Exception as e:
+            logger.error(
+                f"entity_manager | delegate.py | Unable to parse delegate metadata while indexing: {e}"
+            )
+            return None
+    return metadata
+
+
+def validate_delegate_tx(
+    params: ManageEntityParameters, metadata: CreateDelegateMetadata
+):
+    user_id = params.user_id
+
+    if params.entity_type != EntityType.DELEGATE:
+        raise Exception(
+            f"Invalid Delegate Transaction, wrong entity type {params.entity_type}"
+        )
+
+    if params.action == Action.CREATE:
+        if not metadata["address"]:
+            raise Exception(
+                "Invalid Delegate Transaction, address is required and was not provided"
+            )
+        if not metadata["name"]:
+            raise Exception(
+                "Invalid Delegate Transaction, name is required and was not provided"
+            )
+        if metadata["address"].lower() in params.existing_records[EntityType.DELEGATE]:
+            raise Exception(
+                f"Invalid Delegate Transaction, address {metadata['address']} already exists"
+            )
+        if user_id and user_id not in params.existing_records[EntityType.USER]:
+            raise Exception(
+                f"Invalid Delegate Transaction, user id {user_id} does not exist"
+            )
+
+        if metadata["address"].lower() != params.signer.lower():
```

Comment:
i thought the address was the API key - is this right? normally we would compare the signer w the user's wallet. i think for this the user is optional so maybe we verify that if it's via a user logged in via UI?

Reply:
yeah that sgtm! trying to think of any risks to allowing someone to create a delegate under an address that they don't own, but I think it's all good